doc/mgmt/Management Framework.md

```diff
@@ -0,0 +1,2029 @@
+    CreateEntry(k,v)|    none          | HMSET(fields in v)| ValidateEditConfig(
+                    |                  |                   |  OP_CREATE)
+    ----------------|------------------|-------------------|--------------------
+    ModEntry(k,v)   | mod_entry(k,v)   | HMSET(fields in v)| ValidateEditConfig(
+                    |                  |                   |  OP_UPDATE)
+    ----------------|------------------|-------------------|--------------------
+    DeleteEntry(k,v)|set,mod_entry(k,0)| DEL               | ValidateEditConfig(
+                    |                  |                   |  OP_DELETE)
+    ----------------|------------------|-------------------|--------------------
+    DeleteEntryField|    none          | HDEL(fields)      | ValidateEditConfig(
+    (k,v)           |                  |                   |   DEL_FIELDS)
+
+
+##### 3.2.2.7 Transformer
+
+Transformer provides the underlying infrastructure for developers to translate data from YANG to ABNF/Redis schema and vice versa, using YANG extensions annotated to YANG paths to provide translation methods. At run time, the YANG extensions are mapped to an in-memory Transformer Spec that provides two-way mapping between YANG and ABNF/Redis schema for Transformer to perform data translation while processing SET/GET operations.
+
+In case that SONiC YANG modules are used by NBI applications, the Transformer performs 1:1 mapping between a YANG object and a SONiC DB object without a need to write special translation codes. If the openconfig YANGs are used by NBI applications, you may need special handling to translate data between YANG and ABNF schema. In such case, you can annotate YANG extensions and write callbacks to perform translations where required.
+
+For special handling, a developer needs to provide:
+1. An annotation file to define YANG extensions on YANG paths where translation required
+e.g. In openconfig-acl.yang, ACL FORWARDING_ACTION "ACCEPT" mapped to "FORWARD", ACL sequence id ‘1’ mapped to ‘RULE_1’ in Redis DB etc. 
+2. Transformer callbacks to perform translation
+
+###### 3.2.2.7.1 Components
+
+Transformer consists of the following components and data:
+* **Transformer Spec:** a collection of translation hints
+* **Spec Builder:** loads YANG and annotation files to dynamically build YANG schema tree and Transformer Spec.
+* **Transformer Core:** perform main transformer tasks, i.e. encode/decode YGOT, traverse the payload, lookup Transformer spec, call Transformer methods, construct the results, error reporting etc.
+* **Built-in Default Transformer method:** perform static translation
+* **Overloaded Transformer methods:** callback functions invoked by Transformer core to perform complex translation with developer supplied translation logic
+* **Ouput framer:** aggregate the translated pieces returned from default and overloaded methods to construct the output payl
```

Comment:
Could you be a little more specific? How does this framework scale in terms of the number of models/APIs? Do you anticipate any significant impact in terms of latency, memory usage, etc? Is this impact static or is it somehow tied to the schema or the number of APIs or some other metric?

Reply:
All the translation metadata – transformer callbacks or simple hints like field-name change etc. -  defined as YANG extensions across YANG modules are loaded and stored in the in-memory Transformer Spec. As the number of YANG modules requiring translation increases, the size of Transformer Spec grows and consumes more memory. During runtime, if the request URI represents a resource which can span multiple augmented modules with a deep depth, the number of external transformer callbacks involved in data translations between YANG and Redis ABNF schema could increase to incur process/method call overhead and consume more resources. In general, the overall latency and resource consumption is mainly dependent on the number of translation points based on the request URI/payload.